spd/app/backend/controller.py

```diff
@@ -0,0 +1,167 @@
+@handle_errors
+def get_status() -> Status:
+    return run_context_service.get_status()
+
+
+@app.post("/runs/load/{wandb_run_id}")
+@handle_errors
+def load_run(wandb_run_id: str):
+    global ablation_service
+    run_context_service.load_run(f"{ENTITY}/{TRAIN_PROJECT}/{wandb_run_id}")
+
+
+@app.get("/available_prompts")
+@handle_errors
+def get_available_prompts() -> list[AvailablePrompt]:
+    return run_context_service.get_available_prompts()
+
+
+@app.post("/run_prompt/{dataset_index}")
+@handle_errors
+def run_prompt_by_index(dataset_index: int) -> RunResponse:
+    """Run a specific prompt from the dataset by index."""
+    return ablation_service.run_prompt_by_index(dataset_index)
+
+
+@app.post("/ablate_subcomponents")
+@handle_errors
+def ablate_subcomponents(request: SubcomponentAblationRequest) -> SubcomponentAblationResponse:
+    tokens_logits = ablation_service.ablate_subcomponents(
+        request.prompt_id,
+        request.subcomponent_mask,
+    )
+    return SubcomponentAblationResponse(token_logits=tokens_logits)
+
+
+@app.post("/simulate_merge")
+@handle_errors
+def simulate_merge(request: SimulateMergeRequest) -> SimulateMergeResponse:
+    """Simulate merging masks without persisting the result"""
+    return ablation_service.get_merge_l0(
+        prompt_id=request.prompt_id, layer=request.layer, token_indices=request.token_indices
+    )
+
+
+@app.post("/combine_masks")
+@handle_errors
+def combine_masks(request: CombineMasksRequest) -> CombineMasksResponse:
+    mask = ablation_service.create_combined_mask(
+        prompt_id=request.prompt_id,
+        layer=request.layer,
+        token_indices=request.token_indices,
+        description=request.description,
+    )
+
+    ablation_service.save_mask(mask)
+
+    return CombineMasksResponse(
+        mask_id=mask.id,
+        mask=mask.to_dto(),
+    )
+
+
+@app.get("/mask")
+@handle_errors
+def get_masks() -> list[MaskDTO]:
+    return [mo.to_dto() for mo in ablation_service.saved_masks.values()]
+
+
+@app.post("/apply_mask")
+@handle_errors
+def apply_mask_as_ablation(request: ApplyMaskRequest) -> AblationResponse:
+    """Apply a saved mask as an ablation to a specific prompt."""
+    tokens_logits, ablation_effect = ablation_service.run_with_mask(
+        request.prompt_id, request.mask_id
+    )
+    return AblationResponse(token_logits=tokens_logits, ablation_effect=ablation_effect)
+
+
+@app.get("/activation_contexts/subcomponents")
+@handle_errors
+def get_subcomponent_activation_contexts(
+    importance_threshold: float,
+    n_batches: int,
+    batch_size: int,
+    n_tokens_either_side: int,
+) -> ModelActivationContexts:
+    assert (run_context := run_context_service.train_run_context) is not None
+    return get_subcomponents_activation_contexts(
+        run_context,
+        importance_threshold=importance_threshold,
+        n_batches=n_batches,
+        n_tokens_either_side=n_tokens_either_side,
+        batch_size=batch_size,
+    )
```

Comment:
Lol please change the names of these functions. IMO fine to just not have the inner function call at all and just call get_topk_by_subcomponent and map_to_model_ctxs directly.